### PR TITLE
5390 correctly handle gs1 2d data matrix format

### LINKED ISSUE #5390: Correctly handle GS1 2D data matrix format
## What went wrong? 😲

We have created a GS1 Barcode format interpreter based on some miss guided assumptions.

I assumed that the GS1 barcode would contain the data elements separated by `(` characters `)` as it is shown for display purposes.

However, the actual data format is more complicated, for example...

```
2.2.1 Function 1 Symbol Character (FNC1)
By definition in ISO/IEC 16022 GS1 DataMatrix uses a special start sequence to differentiate GS1
DataMatrix from other ISO/IEC Data Matrix symbols. This is achieved by using the Function 1
Symbol Character (FNC1) in the first position of the data encoded. It enables scanners to process
the information according to the GS1 System Rules.
The FNC1 (codeword 232) has two separate uses in GS1 DataMatrix:
■ Start character.
■ Separator character to separate element strings that are not in the predefined list (see
Table 2-3).


2.2.2 Concatenation
Using GS1 DataMatrix, it is possible to concatenate (chain together) discrete element strings
(Application Identifiers (AIs) and their data) into a single symbol. When the element string is of
predefined length, no separator character is required when another element string is added to it.
When the element string is not of predefined length, it must be followed by a separator character
when concatenating more element strings. The separator character may be FNC1 or the control
character <GS> (ASCII value 29 (decimal), 1D (hexadecimal)). FNC1 is the character that has the
codeword value 232. When either FNC1 or <GS> is used as a separator character, the scanner must
transfer the character value as control character <GS>.
A separator character is never required after the last element string encoded in the symbol.
```

A flow diagram for interpreting these barcodes, as well as the field length look up table etc are available here:
https://www.gs1.org/docs/barcodes/GS1_DataMatrix_Guideline.pdf (Page 15)

It turns out we already `knew` some of this and are using a front end library to pull out information such as batch number
`const batch = gs1?.parsedCodeItems?.find(item => item.ai === '10')`
https://www.npmjs.com/package/gs1-barcode-parser-mod

So with could either take the output of that library, perhaps convert to JSON? then send to backend (avoiding the need to do any complex logic in rust (Probalby the easiest option)

Alternatively we could, find or build a rust library to interpret the data.
From a quick look, similar to what I found before, there doesn't appear to be a reasonably full featured rust implementation yet...

## Expected behaviour 🤔

We should be able to handle correctly formatted GS1 Data Matrix format

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Scan a GS1 formatted Data Matrix here (Note: or create our own for testing)

PQS Asset Type: E003/002
Serial Number: S12345678
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S12345678(241)E003/002(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101
image

PQS Asset Type: E004/055
Serial Number: S1
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S1(241)E004/055(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101

image

PQS Asset Type: E004/055
Serial Number S2
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S1(241)E004/055(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101

image

## Your environment 🌱

- Open mSupply Version: 2.4-RC1
- Legacy mSupply Central Server Version:

- Platform:

- Database type:

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #5390

# 👩🏻‍💻 What does this PR do?

I tried to make minimal changes for this, but it should allow us to scan GS1 formatted barcodes and create assets from them. 

## 💌 Any notes for the reviewer?

Maybe/probably should remove the parsing of human readable string? Figured I'd leave it there for now as it might come in handy, and I didn't want to have to re-create data for all my tests cases and aiming for minimal changes here :)

Thoughts on this?

# 🧪 Testing

- [ ] Using either Android (with camera), or Desktop (with 2d Barcode Scanner)
- [ ] Open the cold chain equipment section, and click the `Scan` button (this doesn't show on web)
- [ ] Scan one of the barcodes below
- [ ] You should be prompted to create a new asset
- [ ] If you accept, it should show a new asset with the serial number, asset type, and Warenty date populated
- [ ] The Installation date should default to today
- [ ] Scanning the same barcode again, should navigate you to the right asset (once it has been created)

> NOTE, scanning these 2d Matrixes in github `dark mode` or with a black background doesn't work well...
> NOTE: in general scanning 2d Matrixes seems much less reliable than QR Codes, need to have camera straight on for example

Scan a GS1 formatted Data Matrix here (Note: or create your own for testing)

PQS Asset Type: E003/002
Serial Number: S12345678
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S12345678(241)E003/002(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101
image

PQS Asset Type: E004/055
Serial Number: S1
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S1(241)E004/055(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101

image

PQS Asset Type: E004/055
Serial Number S2
https://bwipjs-api.metafloor.com/?bcid=gs1datamatrix&text=(01)09506000134352(11)241007(21)S1(241)E004/055(3121)820000(3131)670000(3111)630000(90)001(91)241007-310101

image

# 📃 Documentation

- [X] **Part of an epic**: documentation will be completed for the feature as a whole
- [X] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx [hunk 1]
```diff
@@ -34,7 +34,8 @@ export const AddFromScannerButtonComponent = () => {
   const equipmentRoute = RouteBuilder.create(AppRoute.Coldchain).addPart(
     AppRoute.Equipment
   );
-  const { mutateAsync: scanAsset } = useAssets.document.scan();
+  const { mutateAsync: fetchAsset } = useAssets.document.fetch();
+  const { mutateAsync: fetchGS1 } = useAssets.document.gs1();
```
Comment (lache-melvin) [inline client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx:38]: NIT: can we call it `fetchAssetByGS1` here too? This feels like it should fetch GS1 data... maybe even `fetchOrCreateFromGS1`? 🤷‍♀️

File: server/service/src/asset/parse.rs [hunk 2]
```diff
@@ -113,23 +96,11 @@ fn create_draft_asset_from_gs1(
     Ok(asset)
 }
 
-pub fn parse_from_scanned_data(
+pub fn create_from_gs1_data(
```
Comment (lache-melvin) [inline server/service/src/asset/parse.rs:99]: Again, nit but maybe `get_or_create_from_gs1_data` is a little clearer?

File: client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx [hunk 3]
```diff
@@ -64,9 +65,24 @@ export const AddFromScannerButtonComponent = () => {
 
   const handleScanResult = async (result: ScanResult) => {
     if (!!result.content) {
-      const { content } = result;
+      const { gs1 } = result;
 
-      const asset = await scanAsset(content).catch(() => {});
+      if (!gs1) {
+        // try to fetch the asset by id, as it could be an id from our own barcode
```
Comment (lache-melvin) [inline client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx:71]: I like the move of this to the frontend!

File: client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx [hunk 4]
```diff
@@ -64,9 +65,24 @@ export const AddFromScannerButtonComponent = () => {
...
+        const { content } = result;
+        const id = content;
```
Comment (lache-melvin) [inline client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx:73]:
```suggestion
        const { content: id } = result;
```

**Review by lache-melvin — APPROVED** [review]
Working great! Thanks James 🚀 

> Maybe/probably should remove the parsing of human readable string? Figured I'd leave it there for now as it might come in handy, and I didn't want to have to re-create data for all my tests cases and aiming for minimal changes here :)

I feel ok leaving it there, it's pretty isolated and comes in handy for testing if nothing else 😁

**Comment by lache-melvin** [comment]
HAHA kidding wait I built it I didn't test hold fire

**Comment by lache-melvin** [comment]
#exposed

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Alas: 
Screenshot_20241114_135559_Open mSupply

File: client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx (hunk 3 above)
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/ListView/AddFromScannerButton.tsx:71]: It's nice now, as we have the frontend library to do the work to figure out if it's GS1.

**Comment by lache-melvin** [comment]
Ah - I think `parseBarcode` isn't returning anything? This is the `ScanResult` in `handleScanResult`:
Screenshot_20241114_162720_Open mSupply

**Comment by lache-melvin** [comment]
I think I see the issue, doing a quick test build now, will update you :)

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Ok there are two issues:

The `parseResult` being called in the electron scanning code, but not for android: 
https://github.com/msupply-foundation/open-msupply/blob/1d76ccd91bdff7ddb71a344aedadb708c4ab50b5/client/packages/common/src/utils/BarcodeScannerContext.tsx#L173-L190

However, I added the `parseResult` call there, and still the gs1 result of `parseBarcode` is coming back empty... don't have time to investigate further tonight sorry, I think a tomorrow problem 😁

**Comment by jmbrunskill** [comment]
> However, I added the `parseResult` call there, and still the gs1 result of `parseBarcode` is coming back empty... don't have time to investigate further tonight sorry, I think a tomorrow problem 😁

Dang, was hopeful!

**Comment by mark-prins** [comment]
> The parseResult being called in the electron scanning code, but not for android:

I have a vague memory that this was deliberate and was struggling to remember why.
In the history of the file there's this comment (removed as the response to PR feedback):

```
  // a data matrix result from the Capacitor barcode scanner plugin
  // is returned in sections, separated by \x1D and needs to be parsed
  // in sections - otherwise the serial number includes the AI characters
  // for batch and expiry date
```

which makes me think that there is a difference in how the barcode is returned.
You guys are busy, let me have a look this morning and see if I can recall things..

**Comment by lache-melvin** [comment]
Aw epic thanks Mark! Logged an unhandled error from that method this morning saying "not a valid AI", so that sounds about right 🙏

**Comment by jmbrunskill** [comment]
> (removed as the response to PR feedback):

Let's keep this stuff in future!

I guess the question is if it's easy to adapt that library to work?

**Comment by mark-prins** [comment]
Have resolved by replacing the `\x1d` character which the capacitor plugin adds and then parsing the result to get the gs1 content.

It's working nicely now for the test barcodes above.


Couple of observations, which I'll save here since I've gone this far..

Android gives the `\x1d` (or decimal 29) at the start, where the USB scanner gives `\x2e\x10\x02\x00`
and the USB ends with `\x00\x32\x0B` and then zeros after that.

```
Electron: [46,16,3,0,48,49,48,57,53,48,54,48,48,48,49,51,52,51,53,50,49,55,50,50,49,50,50,53,49,48,65,66,67,68,69,70,29,50,49,49,50,51,52,0,50,11,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
Android : [29,48,49,48,57,53,48,54,48,48,48,49,51,52,51,53,50,49,55,50,50,49,50,50,53,49,48,65,66,67,68,69,70,29,50,49,49,50,51,52]
```

**Review by mark-prins — APPROVED** [review]
👍 working nicely for me!

**Comment by jmbrunskill** [comment]
Thanks @mark-prins Appreciate the effort!

**Pull request merged**